### PR TITLE
Create PriorityClassName test

### LINKED ISSUE #82545: PodSpec - new priorityClassName test
Umbrella issue: #82312

/area conformance
/sig testing

### PR DESCRIPTION
**Pull request by BobyMCbobs**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Creates PriorityClassName test

**Which issue(s) this PR fixes**:
Fixes #82545

**Special notes for your reviewer**:
Umbrella issue: #82312

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/area testing

### CODE REVIEW

File: test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -467,6 +468,70 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 			}
 		}
 	})
+	// this test makes sure that Pods with a higher PriorityClass via their PriorityClassName start earlier than Pods with a lower PriorityClassName
+	ginkgo.It("should ensure Pods using a PriorityClassName which is higher begins first", func() {
```
Comment (alejandrox1) [inline test/e2e/scheduling/preemption.go]: Should this be "should ensure Pods using a PriorityClassName which is higher begin first" ?

**Review by alejandrox1 — COMMENTED** [review]
/ok-to-test
/priority backlog
/sig scheduling

File: test/e2e/scheduling/preemption.go (hunk 1 above)
Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: Thanks for the suggestion, I'll update that :)

Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: Updated testname in #50de4bfa9e52d2fba96458a88f093a30606db892

File: test/e2e/scheduling/preemption.go [hunk 2]
```diff
@@ -467,6 +468,70 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
...
+		defer close(stopCh)
+
+		ginkgo.By("Creating Pods with priority")
+		for podNum := 0; podNum < len(pods); podNum++ {
+			podName := pods[podNum].ObjectMeta.Name
+			fmt.Println(fmt.Sprintf("Creating Pod %v/%v: %v", podNum+1, len(pods), podName))
+			_, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(&pods[podNum])
+			framework.ExpectNoError(err, "Failed creating Pod")
+		}
+		fmt.Println(fmt.Sprintf("Pods seen so far: %v", podNamesSeen))
```
Comment (draveness) [inline test/e2e/scheduling/preemption.go]: why not use `e2elog` here?

Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: Thanks for pointing this out.

I've updated it use use `e2elog.Logf` instead, in fd0e5475c5e0173db27ceebb85f34bc5b87dabc7

File: test/e2e/scheduling/preemption.go [hunk 3]
```diff
@@ -467,6 +468,71 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 			}
 		}
 	})
+	// this test makes sure that Pods with a higher PriorityClass via their PriorityClassName start earlier than Pods with a lower PriorityClassName
```
Comment (Huang-Wei) [inline test/e2e/scheduling/preemption.go]: This statement doesn't look right. PriorityClass only takes into effect when the incoming pod can't fit on any node - when the so-called "preemption" happens.

File: test/e2e/scheduling/preemption.go [hunk 4]
```diff
@@ -467,6 +468,71 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
...
+		defer close(stopCh)
+
+		ginkgo.By("Creating Pods with priority")
+		for podNum := 0; podNum < len(pods); podNum++ {
+			podName := pods[podNum].ObjectMeta.Name
+			e2elog.Logf("Creating Pod %v/%v: %v", podNum+1, len(pods), podName)
+			_, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(&pods[podNum])
+			framework.ExpectNoError(err, "Failed creating Pod")
+		}
+		//fmt.Println(fmt.Sprintf("Pods seen so far: %v", podNamesSeen))
```
Comment (draveness) [inline test/e2e/scheduling/preemption.go]: This line shoud be removed

**Review by draveness — CHANGES_REQUESTED** [review]

File: test/e2e/scheduling/preemption.go [hunk 5]
```diff
@@ -467,6 +468,71 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
...
+		ginkgo.By("Defining Pod specs")
+		for i := 4; i >= 1; i-- {
+			pods = append(pods, v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("priority-class-p%v", i),
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:    "priority-class",
+						Image:   imageutils.GetE2EImage(imageutils.Agnhost),
```
Comment (Huang-Wei) [inline test/e2e/scheduling/preemption.go]: Why using this image?

File: test/e2e/scheduling/preemption.go (hunk 4 above)
Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: Whoops! That wasn't meant to be in there.

Thanks for finding it, I've removed it in https://github.com/kubernetes/kubernetes/pull/82546/commits/3e5d4a41d5f29a03546a76ffcb4c4f181e383763

File: test/e2e/scheduling/preemption.go [hunk 6]
```diff
@@ -467,6 +468,71 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
...
+		stopCh := make(chan struct{})
+		pods := []v1.Pod{}
+		// define Pod specs - order in from highest to lowest priority
+		ginkgo.By("Defining Pod specs")
+		for i := 4; i >= 1; i-- {
+			pods = append(pods, v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("priority-class-p%v", i),
+				},
+				Spec: v1.PodSpec{
```
Comment (Huang-Wei) [inline test/e2e/scheduling/preemption.go]: The PodSpec doesn't claim any type of resource usage. So they won't test preemption at all.

**Review by Huang-Wei — CHANGES_REQUESTED** [review]
The statement of "ensure Pods using a PriorityClassName which is higher begin first" isn't true. Please clarify what you want to test, then I can help to come up with a 
meaningful test case.

File: test/e2e/scheduling/preemption.go (hunk 5 above)
Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: I chose this image because it's platform agnostic

File: test/e2e/scheduling/preemption.go (hunk 6 above)
Comment (BobyMCbobs) [reply test/e2e/scheduling/preemption.go]: OK, is there a resource which would be recommended to use then?

**Comment by Huang-Wei** [comment]
@BobyMCbobs @alejandrox1 May I know in which aspect the `spec.priorityClassName` should be covered in a conformance test?

**Comment by alejandrox1** [comment]
Thank you for your comments @Huang-Wei !
From the looks of https://github.com/kubernetes/kubernetes/issues/82312 it seems that we just want to test the "general" behavior of pod priority and preemption.
Specifically [schedulingorder](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order):
> scheduler orders pending Pods by their priority and a pending Pod is placed ahead of other pending Pods with lower priority in the scheduling queue. As a result, the higher priority Pod may be scheduled sooner than Pods with lower priority if its scheduling requirements are met. If such Pod cannot be scheduled, scheduler will continue and tries to schedule other lower priority Pods.

So to make this test valid, the pods being created here should be first marked as pending (not enough space for them to run in the cluster).
D you know of a good way of doing this @Huang-Wei or of a better alternative ?

**Comment by Huang-Wei** [comment]
@alejandrox1 Ah I see why this case is composed. Actually, only upon a high-churn & high-utilized cluster - in other words, a lot of pods are pending in the internal queue - pods with high priority get picked up by scheduler earlier than low priority ones even if the high priority one is created after low priority one. So it's possible to simulate this in a unit test/integration test, but will be a little challenging to simulate in an e2e test.

Since this is an internal behavior/implementation, I suppose it doesn't reach the bar of being a conformance test. Any thoughts?

**Comment by devaips** [comment]
> Since this is an internal behavior/implementation, I suppose it doesn't reach the bar of being a conformance test. Any thoughts?

Not sure what you mean here @Huang-Wei. Is this something that can be mapped to the [conformance test requirements doc?](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements)

**Comment by Huang-Wei** [comment]
@devaips I mean that this test _tries_ to test a very internal behavior which shouldn't be a conformance test. And no need to mention this PR can't verify the internal behavior either - as I mentioned earlier.

**Comment by hh** [comment]
This is a ping for 'is this a conformance test' feedback:
/assign @johnbelamaric @timothysc

File: test/e2e/scheduling/preemption.go (hunk 3 above)
Comment (johnbelamaric) [reply test/e2e/scheduling/preemption.go]: https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/

File: test/e2e/scheduling/preemption.go (hunk 6 above)
Comment (johnbelamaric) [reply test/e2e/scheduling/preemption.go]: the point here is you need to specify cpu or memory requests. For example: https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/#specify-a-memory-request-and-a-memory-limit

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
review the docs on the functionality and revise. thanks!

File: test/e2e/scheduling/preemption.go (hunk 5 above)
Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: These are the defaults now and the correct images to use.

**Review by timothysc — COMMENTED** [review]
So the test has some edges.  It's totally possible to hit a condition where you can hit a scheduling run in the middle.  The behavior test should really be about preemption as that is the intended consequence for this feature.  Although scheduling order is useful it's not the primary use case.

**Comment by Huang-Wei** [comment]
This PR basically creates 4 pods in batch, with the order P4 (with priority value 4), P3, P2 and P1, and then verifies if the informer receives the 4 pods in the exact order. This flow can't prove how PriorityClass works or not - b/c no preemption happens here, hence you always get the Pod created earlier than the one created later, no matter what their PriorityClass is.

If you really want to verify that high priority Pod will suppress the "preemption" ability of lower Pod, test [PreemptionExecutionPath] does the job.

So I'm leaning towards closing this issue, and if appropriate, promote [PreemptionExecutionPath] test to a conformance test.

**Comment by spiffxp** [comment]
Discussed during conformance meeting. Why was this test written? As in, what specific endpoints was it written to hit that aren't already hit?

There is a desire where possible when writing tests exercising new API endpoints not just to verify CRUD functionality, but also the intended effect.  If the PreemptionExecutionPath test does this already, we should close this.  If there is a less circuitous way of verifying PriorityClass endpoints, we should do that here.

**Comment by hh** [comment]
In September, when we created this was before #84725 merged.

**Pull request closed without merging**